### ISSUE TITLE
Encryption of GnuPG messages should omit certain keying information

### ISSUE THREAD

**Issue by ioerror** [freedomofpress/securedrop#13]
I'm skeptical of using GnuPG but if you're going to use it, I suggest you blind the messages such that the key ID is not embedded - this should at least ensure that a specific key may not be demanded by cryptographic identifier :

```
       --hidden-recipient name

       -R     Encrypt for user ID name, but hide the key  ID  of  this  user's
              key.  This  option helps to hide the receiver of the message and
              is a limited countermeasure against traffic  analysis.  If  this
              option  or --recipient is not specified, GnuPG asks for the user
              ID unless --default-recipient is given.
```

There are other reasons but the above seems sufficient to me.

**Comment by garrettr** [comment]
I don't think hiding the recipient is a useful protection given SecureDrop's current architecture. The primary use seems to be in the PGP-email use case, where PGP encrypted messages are often transferred over an unencrypted transport. An adversary that intercepts them may use the recipient keyid to infer metadata and build a social network graph.

In SecureDrop, all connections made to the Application Server (by both sources and journalists) are via the Tor Hidden Service protocol, so they are end-to-end encrypted. Additionally, there is only one direction (source -> journalist) where the actual GPG messages are transmitted, and in that case the recipient is obviously the journalist. Further, at the moment that there is only one "application key" used by SecureDrop, since we haven't implemented multiple recipients support yet. Therefore, even if you could obtain the GPG messages (by hacking the Application Server or breaking the transport encryption for THS as the journalist downloads submissions), you would not learn any new information from the recipient metadata because the message is obviously encrypted to the single "application key".

Implementing this would have value *only* if we implemented either of the two features (which have been bandied about in various conversations):

1. Allow sources to upload their own GPG key for SecureDrop to use in encrypting the replies sent to them. This would be especially important because the replies are encrypted with *both* the source's key and the journalist's key (so the journalist can read it later), and if the source had these messages on their machine the presence of the journalist's keyid would be a conclusive link to their use of SecureDrop.
2. Implemented multiple recipient support, in which case it might useful as defense in depth

**Comment by garrettr** [comment]
Closing this for now. It should be re-opened and resolved as part of either of the linked issues (#144 or #98).

**Comment by ninavizz** [comment]
If it would not ever compromise security to expose the recipient's name, I see this as an epic usability flag—and it should get studied in RITE testing sessions w/ a prototype that'd mimic implied backend behaviors, before worked-on as a security feature for a release.

**Comment by fowlslegs** [comment]
@ninavizz I don't understand what you mean when you say:

> If it would not ever compromise security to expose the recipient's name

Can you rephrase?

Also, what's RITE?

**Comment by ninavizz** [comment]
RITE testing is the only form of research I really get behind—the academic whitepaper bullshit has accomplished little more than expensive bellybutton lint rearrangement, for all of the hundreds of thousands of dollars thrown at it on infosec matters. RITE testing requires competency in both design and research, and is fun as shit. Rapid Iterative Testing Evaluation. Basically: design, prototype, test (with users); whiteboard findings, brainstorm, fix design, fix prototype, re-test; rinse, wash, repeat! Way the fuck quicker, no pretentious career-advancing hooey, and users reap the rewards—not researcher egos. The Journalist and Admin interfaces imho, could benefit a lot from it!

http://www.answerlab.com/blog/2013/10/22/rite-when-you-need-to-know-fast/

re: recipients name thing... I possibly don't understand what Jake and Garrett are blabbing about, but my read on it is that Jake suggested (and then Garrett clarified in his last point in his last comment) that with multiple journalist keys it somehow may benefit user (or system) security to hide the journalist's name in transmission, when a source chooses to send their information to a specific journalist/journalist-key. I can see how it could compromise things should a hacker intercept a leak to Jane Journalist, who could then get subpoenaed or stalked or whatnot by association, but I'm curious about what user-facing ramifications of such a thing would be. Plainly: In a UI, would users have to compensate for an absent human's name as a visible recipient to sent things? 

I'm only caring, because the primary knock against SD vs GL right now (that I've read at least), is that SD is undesirable for both sources and journalists, because its usability sucks—while it's also too cumbersome to instal/maintain. Sucky usability is much easier to correct than sucky security—and GL has intentionally chosen the more lightweight security route, I've read, to be easier to install/maintain—but I don't see why there should have to be such a compromise for one or the other. Users are the ones who ultimately suffer either, and usability compromises simply not ever needing to exist imho, I have so little patience for them. So much so, that I'm willing to drive knowledgeable developers bonkers with my longwinded misunderstandings of their problems, to find opportunities y'all may not be seeing. So—if that happens to be the case here, feel free to ignore me, just please NEVER accept a security best practice as mandating a usability compromise. Opportunities for usability are never hopeless, unless a designer says so (and usability fucking matters, dammit)! </lecture>
